[feat][#48] 모임원 초대 링크 생성 API 구현

h2jinee opened the pull request:
## 관련 이슈
- resolve #48 

## 작업 내용
- [x] MeetingDto 관련 누락된 수정사항 반영
  - MeetingStatus enum으로 변경된 `status` 필드의 테스트 코드 미적용 부분 수정
  - `UpdateStatus` API 응답 구조를 Response 객체를 사용하도록 변경
  - `CommonResponse`를 sealed class로 변환하여 Success/Error 타입 분리
- [x] 엔티티 코드 컨벤션 적용
  - `Meeting`, `Member` 엔티티에서 누락된 companion object logger 패턴 적용
  - `GlobalExceptionHandler`의 logger 변수명을 프로젝트 컨벤션에 맞게 log로 통일
  - 모든 엔티티의 시간 타입을 `LocalDateTime`에서 `Instant`로 변경
- [x] 모임 초대 링크 기능 구현
  - `invite_links` 테이블 생성 및 관련 인덱스 추가
  - 초대 링크 생성 API 구현 (`POST /api/v1/meeting/{id}/invite`)
  - 모임 멤버 수 제한(5명) 검증 로직 추가
  - 초대 링크 만료 시간 관리 구현
- [x] 환경 설정 개선
  - Swagger local 환경 설정으로 통일
  - 초대 링크 `base-url` 설정 추가

## 이슈 및 해결 과정
1. **초대 링크 생성 방식 검토**
    - **원인 분석**:
        1. Firebase Dynamic Links를 활용하여 초대 기능을 구현하려 했으나, 해당 서비스가 곧 종료될 예정인 것을 알게 되었습니다.
        2. 다른 대체 서비스들을 검토했으나, 비용 문제가 있었습니다.
        
    - **해결 방법**:
        1. 우선 자체 초대 링크 시스템을 구축하기로 결정했습니다.
        2. `invite_links` 테이블을 설계하고 초대 코드 생성 로직을 구현했습니다.
        
            ```sql
            CREATE TABLE invite_links
            (
                id         BIGINT      NOT NULL AUTO_INCREMENT COMMENT '초대 링크 ID',
                meeting_id UUID        NOT NULL COMMENT '모임 ID',
                code       VARCHAR(36) NOT NULL COMMENT '초대 코드',
                creator_id UUID        NOT NULL COMMENT '생성자 ID',
                expired_at DATETIME(6) NOT NULL COMMENT '만료일시',
                created_at DATETIME(6) NOT NULL DEFAULT CURRENT_TIMESTAMP(6) COMMENT '생성일시',
                updated_at DATETIME(6) NOT NULL DEFAULT CURRENT_TIMESTAMP(6) COMMENT '수정일시',
                PRIMARY KEY (id),
                CONSTRAINT uk_invite_links_code UNIQUE (code),
                CONSTRAINT fk_invite_links_meeting FOREIGN KEY (meeting_id) REFERENCES meeting (id),
                CONSTRAINT fk_invite_links_creator FOREIGN KEY (creator_id) REFERENCES user (id)
            ) COMMENT '초대 링크';
            ```

2. **시간 필드 타입 불일치**
    - **원인 분석**:
        1. `LocalDateTime` 사용으로 인해 서버 간 시간 동기화 문제가 발생했습니다.
        2. 글로벌 서비스 확장 시 시간대 이슈가 예상되었습니다.
        
    - **해결 방법**:
        1. 모든 엔티티의 시간 관련 필드를 `Instant` 타입으로 변경했습니다.
        2. UTC 기준으로 시간을 저장하여 시간대 불일치 문제를 해결했습니다.

3. **API 응답 구조 문제**
    - **원인 분석**:
        1. `CommonResponse`가 단순히 status와 message만을 포함하고 있어 응답 데이터의 타입 안전성이 보장되지 않았습니다.
        2. 모임 상태 변경 API는 다른 API들과 달리 Response 객체를 사용하지 않아 일관성이 떨어졌습니다.
        
    - **해결 방법**:
        1. `CommonResponse`를 sealed class로 재설계했습니다.
        
            ```kotlin
            sealed class CommonResponse<T> {
                data class Success<T>(val data: T) : CommonResponse<T>()
                data class Error(val status: Int, val message: String) : CommonResponse<Nothing>()
            }
            ```
        2.모임 상태 변경 API의 응답도 Response 객체를 사용하도록 수정했습니다.

4. **Swagger 환경 설정 오류**
    - **원인 분석**:
        1. local profile 활성화 시 `swagger.dev` 프로퍼티를 찾지 못해 `SwaggerConfig`의 `@Value`에서 오류가 발생했습니다.
        
    - **해결 방법**:
        1. 우선 `SwaggerConfig`에 local 환경 설정만 남기고 dev 관련 설정을 제거했습니다.
        
            ```kotlin
            class SwaggerConfig(
                @Value("\${swagger.local.url}") private val localUrl: String,
                @Value("\${swagger.local.description}") private val localDescription: String,
            ) {
            ```
                    
        2. `application.yml`에 초대 링크 `base-url` 설정을 추가했습니다.
        
            ```yaml
            app:
              meeting:
                invite:
                  base-url: "http://localhost:8080"
            ```

## 테스트 방법
1. **모임 상태 변경 API 테스트**
    - `PATCH /api/v1/meeting/{id}/status` 엔드포인트가 정상적으로 동작하는지 확인합니다.
    - 변경된 Response 객체 구조가 올바르게 반환되는지 검증합니다.
      - 요청 데이터 예:
      
        ```json
        {
          "status": "ENDED"
        }
        ```
        
      - 예상 응답 데이터 예:
      
        ```json
        {
          "id": "bc0de3e8-d0e5-11ef-97fd-2cf05d34818a",
          "status": "ENDED"
        }
        ```

2. **초대 링크 생성 API 테스트**
    - `POST /api/v1/meeting/{id}/invite` 요청 시 정상적으로 초대 링크가 생성되는지 확인합니다.
    - 생성된 초대 코드로 `GET /api/v1/invite/{code}` 요청을 보내 유효성을 검증합니다.
    - 응답 예시:
    
        ```json
        {
          "inviteUrl": "http://localhost:8080/meeting/invite/abcdef123456",
          "expiredAt": "2024-02-16T10:00:00Z"
        }
        ```

3. **단위 테스트 실행**
    - `MeetingControllerTest`에서 다음 사항들을 검증합니다:
        - `MeetingStatus` enum 타입 변환이 정상적으로 동작하는지 확인합니다.
        - 모임 상태 변경 API가 변경된 응답 구조로 올바르게 응답하는지 확인합니다.
        - 초대 링크 생성의 성공/실패 케이스가 의도한 대로 동작하는지 확인합니다.

## 참고 사항
- 초대 링크는 생성 시점으로부터 3일간 유효합니다.
- SwaggerConfig 환경 설정은 추후 환경별(local, dev, prod)로 개선할 예정입니다.

h2jinee committed:
[Fix] Meeting DTO의 status 필드를 MeetingStatus enum 타입으로 수정하여 타입 에러 해결 (#48)

h2jinee committed:
[Fix] Meeting, Member 엔티티의 logger를 companion object로 변경 (#48)

h2jinee committed:
[Feat] 모임 초대 링크 테이블 및 인덱스 추가 (#48)

h2jinee committed:
[Refactor] CommonResponse를 sealed class로 변환 및 Success/Error 상태 추가로 타입 안전성 강화 (#48)

h2jinee committed:
[Fix] SwaggerConfig dev 환경 설정 제거(임시) 및 불필요한 프로퍼티 정리 (#48)

h2jinee committed:
[Refactor] 로깅 변수 명칭을 log로 통일 (#48)

h2jinee committed:
[Feat] 모임원 초대 링크 생성 API 구현 (#48)

Commit:
[Refactor] 모임 초대 코드 생성 API 재시도 로직 제거 및 단순화 (#48)

Commit:
[Refactor] 엔티티 시간 타입 LocalDateTime -> Instant로 통일 (#48)

h2jinee committed:
[Refactor] Instant로 변경되지 않은 부분 수정 (#48)

f-lab-saponin commented on .github/workflows/deploy.yml:
```diff
@@ -64,4 +64,5 @@ jobs:
             -e LOGGING_LEVEL_ORG_HIBERNATE_SQL=DEBUG \
             -e SWAGGER_DEV_URL=http://175.45.192.193 \
             -e SWAGGER_DEV_DESCRIPTION="Development Server" \
+            -e APP_MEETING_INVITE_BASE_URL=https://ring-go.dev \
```
프로파일 설정 별로 다른 base_url 을 원하신다면 application.yaml 에서도 가능할 것 같아요.

h2jinee replied:
감사합니다! application-dev.yml에 base_url을 추가하고, deploy.yml에 `SPRING_PROFILES_ACTIVE`를 사용하는 것으로 변경했습니다.

f-lab-saponin commented on src/main/kotlin/com/ringgo/common/dto/CommonResponse.kt:
```diff
@@ -1,12 +1,19 @@
-    val status: Int,
-    val message: String
-) {
+sealed class CommonResponse<T> {
+    data class Success<T>(
+        val data: T
+    ) : CommonResponse<T>()
+
+    data class Error(
+        val status: Int,
+        val message: String
+    ) : CommonResponse<Nothing>()
```
👍

f-lab-saponin commented on src/main/kotlin/com/ringgo/common/exception/GlobalExceptionHandler.kt:
```diff
@@ -11,7 +11,7 @@ import org.springframework.web.bind.annotation.RestControllerAdvice
 
     companion object {
-        private val logger = KotlinLogging.logger {}
+        private val log = KotlinLogging.logger {}
```
이건 클래스 바깥쪽 영역에 최상단에 설정하기도 하는데요. companion object 에 선언하는것과 어떤 차이가있을까요?

h2jinee replied:
파일 최상단에 선언하면 해당 파일 내에서만 접근 가능한 프로퍼티가 되고 파일이 로드될 때 초기화됩니다. 그리고 companion object와 달리 JVM 상에서 추가적인 객체가 생성되지 않습니다. companion object 내부에 선언하면 static 멤버처럼 동작하고 클래스 로드 시 초기화됩니다. JVM에서는 companion object가 별도의 내부 클래스로 컴파일되므로 추가적인 객체가 생성됩니다.

f-lab-saponin replied:
그런장점이 있다면 파일 상단으로 옮겨봐도 좋을 것 같아요. 🙂
라이브러리 설명에도 그런 예시를 주고 있네요
https://github.com/oshai/kotlin-logging?tab=readme-ov-file#getting-started

h2jinee replied:
넵! 수정했습니다 😊 감사합니다!

f-lab-saponin commented on src/main/kotlin/com/ringgo/domain/meeting/dto/MeetingDto.kt:
```diff
@@ -42,7 +43,7 @@ class MeetingDto {
             val status: MeetingStatus,
             val memberCount: Long,
-            val createdAt: String,
+            val createdAt: Instant,
```
Instant 로 하면 혹시 시간을 입력할 때는 어떻게 집어넣나요?
추가로 이거 사용하는 테스트쪽에서 스트링으로 그대로 되어있는데 수정한번 부탁드려요!

h2jinee replied:
Instant 타입에 시간을 입력할 때는 ISO 8601 형식의 문자열을 사용합니다. "2024-03-01T12:34:56Z"와 같이 입력하면 Instant.parse("2025-02-18T23:34:56Z")를 통해 Instant 객체로 생성할 수 있는데요! 여기서 Z는 UTC 표준시를 의미합니다. Spring Boot와 Jackson은 Java Time API를 지원하는 모듈(`jackson-datatype-jsr310`)을 기본적으로 포함하고 있어서 JSON으로 ISO 8601 형식의 문자열을 전달하면 자동으로 Instant로 변환합니다.

f-lab-saponin commented on src/main/kotlin/com/ringgo/domain/meeting/dto/MeetingWithMemberCount.kt:
```diff
@@ -12,5 +13,5 @@ data class MeetingWithMemberCount(
     val memberCount: Long,
     val creator: User,
-    val createdAt: LocalDateTime
+    val createdAt: Instant,
```
LocalDateTime 을 사용하고, JVM 옵션을 넣어서 서버의 시간대를 한국시간(UTC+0900) 으로 고정하여 실행하는 방법도 있습니다. 
한번 장단점을 비교해볼 수 있을까요?

h2jinee replied:
Instant 사용 방식은 UTC를 기준으로 계산되므로 시간대에 영향을 받지 않고, 전 세계적으로 동일한 기준 시간을 사용하므로, 시간 계산의 일관성을 제공한다는 장점이 있습니다. 하지만 사용자에게 보여줄 때에는 유저의 로컬 시간에 맞게 항상 변환 과정이 필요하다는 단점이 있습니다.

LocalDateTime + JVM 옵션 (UTC+0900)을 사용하게 되면 데이터베이스에 저장된 시간이 바로 한국 시간이라 조회 시 이해하기 쉽고 시간대 변환 없이 한국 시간을 직관적으로 다룰 수 있어 코드에서도 별도의 시간대 변환 로직이 필요 하지 않습니다. 하지만 서버가 다른 리전으로 이전하거나 글로벌 서비스로 확장 시 문제가 발생할 수 있습니다!

f-lab-saponin replied:
잘 찾아주셔서 감사합니다 👍
그럼 Instant 로 하실 때 JPA 와도 잘 연동되어 DB 에 저장되는 것 까지만 확인부탁드립니다!

h2jinee replied:
DB에 저장된 값을 확인해보니 KST로 저장되고 있었네요..! 스키마의 DATETIME을 TIMESTAMP로 변경했습니다! application.yml도 같이 hibernate.jdbc.time_zone=UTC 설정을 추가했습니다. `java.time.Instant`는 JPA/Hibernate에서 기본적으로 SQL의 TIMESTAMP 타입에 매핑돼서 h2 콘솔 대신 `println`로 UTC 시간으로 저장되는 것도 확인했습니다!

f-lab-saponin commented on src/main/kotlin/com/ringgo/domain/meeting/entity/MeetingInvite.kt:
```diff
@@ -0,0 +1,79 @@
+            )
+        }
+
+        private fun validateMemberLimit(meeting: Meeting, memberRepository: MemberRepository) {
+            val memberCount = memberRepository.countByMeetingId(meeting.id)
```
repository 를 domain 객체 안으로 의존성을 가져오는 것은 주로 사용하지 않는 패턴인데 왜 그런지 생각해볼 수 있을까요?

h2jinee replied:
레이어드 아키텍처에서 Entity는 데이터 구조와 기본적인 비즈니스 규칙을 담당하고 Repository는 데이터베이스 접근과 관련된 책임을 담당하는데요. Entity가 Repository를 직접 사용하면 하나의 클래스가 두 가지 책임을 갖게 되어 SRP 원칙을 위반하게 됩니다. 그리고 이 프로젝트의 경우 Service -> Repository -> Entity 형태의 의존성 흐름을 가지고 있는데, Entity가 Repository를 의존하게 되어 DIP 원칙도 위반하게 됩니다.
회원 수 제한 검증 로직을 Service 계층에서 처리하도록 수정했습니다! 감사합니다. 😊

f-lab-saponin commented on src/main/kotlin/com/ringgo/domain/meeting/entity/MeetingInvite.kt:
```diff
@@ -0,0 +1,79 @@
+            }
+        }
+
+        private fun generateBaseCode() =
+            (UUID.randomUUID().toString() + System.nanoTime())
+                .replace("-", "")
+                .take(CODE_LENGTH)
+
+        private fun generateUniqueCode(base: String, attempt: Int = 0): String {
+            val suffix = if (attempt > 0) attempt.toString() else ""
+            return base.take(CODE_LENGTH - suffix.length) + suffix
+        }
```
code 를 String 대신 Code 라는 값객체로 만들어보고 그 안에서 하면 좀 더 응집력이 생길 수 있을 것 같아요!

h2jinee replied:
수정했습니다! 감사합니다 😊

f-lab-saponin commented on src/main/kotlin/com/ringgo/domain/meeting/service/MeetingInviteService.kt:
```diff
@@ -0,0 +1,44 @@
+    private val memberRepository: MemberRepository,
+    private val meetingInviteConfig: MeetingInviteConfig
+) {
+    @Transactional
+    fun createInviteLink(id: UUID, user: User): MeetingDto.InviteLink.CreateResponse {
+        val meeting = meetingRepository.findByIdOrNull(id)
+            ?: throw BusinessException(ErrorCode.MEETING_NOT_FOUND)
+
+        val invite = meetingInviteRepository.save(
+            MeetingInvite.create(
+                meeting = meeting,
+                creator = user,
+                expirationDays = meetingInviteConfig.inviteExpirationDays,
+                memberRepository = memberRepository
+            )
+        )
+
+        return MeetingDto.InviteLink.CreateResponse(
+            inviteUrl = "${meetingInviteConfig.baseUrl}/meeting/invite/${invite.code}",
+            expiredAt = DateTimeFormatter.ISO_INSTANT.format(invite.expiredAt)
+        )
```
트랜잭셔널 어노테이션을 붙힐 때는 여기에 정말 트랜잭션이 필요한가를 생각해보고 넣으면 좋을 것같아요.
여기서는 한 트랜잭션 안에서 어떤 작업들이 수행되어야 하나요?

h2jinee replied:
이 메서드에서 수행되는 작업은
1. 모임이 실제로 존재하는지 확인합니다. (읽기)
2. 현재 모임의 멤버 수가 최대 허용 인원(5명)을 초과하지 않는지 확인합니다. (읽기)
3. 초대 링크를 생성합니다. (쓰기)

초대 링크 생성 시점에 이미 가득 찬 모임에 대해 불필요한 초대장이 생성되는 것을 방지하기 위해 `@Transactional`를 선언했는데요! 다시 생각해보니 동시성 이슈는 실제 가입 시점에 중요하니 `joinWithInviteCode`에서만 선언하는 것이 적절할 것 같습니다! 수정했습니다. 감사합니다!

f-lab-saponin replied:
동시성 이슈를 말씀하셨는데, joinWithInviteCode 에서 트랜잭션을 통해 이를 해결할 수 있다는 의미이신가요?
트랜잭션은 쓰기연산들의 원자성을 보장하기 위해 사용되는데요. 그렇게 생각한 이유가 궁금합니다!

h2jinee replied:
앗 죄송합니다 오타였습니다 🥹 동시성이 아니라 원자성입니다.. 초대 링크 생성은 다른 작업과 원자성을 보장할 필요가 없고 초대 링크가 생성되더라도 실제 가입은 `joinWithInviteCode`에서 이루어기 때문에 가입 시점에 원자성이 중요합니다.

f-lab-saponin commented on src/main/resources/db/schema.sql:
```diff
@@ -213,22 +213,40 @@ CREATE TABLE vote_participation
-CREATE INDEX idx_activity_meeting ON activity (meeting_id);
\ No newline at end of file
+CREATE INDEX idx_activity_meeting ON activity (meeting_id);
+CREATE INDEX idx_invite_links_expired_at ON invite_links (expired_at);
```
이 인덱스는 왜 필요한가요~?

h2jinee replied:
현재 activity 테이블의 type이 지출 하나만 존재하고, invite_links는 code로만 조회하는 패턴이어서 두 인덱스 모두 필요하지 않을 것 같습니다. 인덱스 제거하고 push 완료했습니다. 감사합니다!!

f-lab-saponin commented on src/main/kotlin/com/ringgo/domain/meeting/entity/MeetingInvite.kt:
```diff
@@ -0,0 +1,79 @@
+import java.util.*
+
+@Entity
+@Table(name = "invite_links")
+@EntityListeners(AuditingEntityListener::class)
+class MeetingInvite(
```
테이블명과 엔티티명이 동일한 규칙으로 만들어졌으면 좋겠는데 어떠신가요?

h2jinee replied:
다른 테이블명들과 일관성 있게 meeting_invite로 테이블명을 바꾸는 게 좋을 것 같아 수정했습니다. 감사합니다!

f-lab-saponin commented on src/test/kotlin/com/ringgo/domain/meeting/controller/MeetingControllerTest.kt:
```diff
@@ -171,4 +176,73 @@ class MeetingControllerTest {
+            // given
+            val expectedResponse = MeetingDto.InviteLink.CreateResponse(
+                inviteUrl = "http://localhost:8080/invite/test-code",
+                expiredAt = "2025-02-09T10:00:00"
```
여기 수정이 필요해보입니다! 😀

h2jinee replied:
baseUrl을 @Value로 주입받도록 변경하고 날짜 포맷을 ISO 8601 형식으로 수정했습니다. 감사합니다!

f-lab-saponin reviewed (changes requested):
모임원 초대링크 만드시느라 고민이 많으셨을 것 같아요!
고생하셨습니다.
코멘트 확인한번 보시고 반영 부탁드립니다! 🙂

h2jinee committed:
[feat][#49] 모임원 초대 수락 API 구현 (#112)

* [Feat] 모임원 초대 수락 API 구현 (#49)

* [Fix] 모임원 초대 링크 생성 API와 충돌 해결 (#49)

* [feat][#50] 모임원 목록 조회 API 구현 (#115)

* [Fix] MeetingControllerTest의 시간 타입을 String에서 Instant로 수정 (#49)

* [Refactor] BusinessException을 ApplicationException으로 이름 변경 (#50)

* [Feat] 모임원 목록 조회 API 구현 (#50)

* [feat][#51] 모임원 내보내기 API 구현 (#114)

* [Feat] 모임원 내보내기 API 구현 (#51)

* [Fix] 모임원 내보내기 API void 응답으로 수정 (#51)

* [Feat] 모임원 내보내기 API 구현 (#51)

* [Feat] 모임원 내보내기 API 구현 (#51)

* [Fix] Swagger 설정 구조 개선 (#51)

* [Refactor] 모임원 내보내기 API를 PATCH 메서드로 변경 및 모임원 테이블에 status 컬럼 추가 (#51)

* [Refactor] 모임원 내보내기 API에서 요청 body 제거 (#51)

* [Fix] 모임원 내보내기 API 테스트 수정 (#51)

* [Refactor] 모임원 내보내기 API DELETE 메소드로 변경 (#51)

---------

Co-authored-by: h2jinee <h2jinee@gmail.com>

---------

Co-authored-by: h2jinee <h2jinee@gmail.com>

* [Refactor] MeetingInvite 유효성 검사 역할 부여 (#49)

---------

Co-authored-by: h2jinee <h2jinee@gmail.com>

h2jinee committed:
[Refactor] 모임 초대 서비스 검증 로직 개선 (#48)

- checkMemberLimit와 validateMemberLimit 중복 제거
- 검증 메서드 네이밍 통일 (validate prefix 사용)
- checkExistingMembership -> validateExistingMember로 이름 변경하여 명확성 개선

Commit:
[Chore] GitHub Actions 워크플로우 버전 업데이트 (#48)

- actions/checkout을 v3에서 v4로 업데이트
- SonarCloud 액션을 구체적 버전(v4.1.0)으로 지정

Commit:
[Chore] 환경 설정 관리 방식 개선 (#48)

- BASE_URL application-dev.yml에서 관리하는 것으로 변경
- deploy.yml 버전 상향

Commit:
[Chore] 불필요한 인덱스 제거 (#48)

- activity.meeting_id 인덱스 제거 (현재 활동 유형이 지출 하나뿐이라 불필요)
- invite_links.expired_at 인덱스 제거 (code로만 조회하는 현재 사용 패턴에 불필요)

Commit:
[Test] MeetingController 초대 링크 생성 테스트 개선 (#48)

- baseUrl 설정값을 @Value 어노테이션으로 주입하도록 수정
- 초대 코드 생성 방식을 실제 구현과 유사하게 변경
- 테스트 데이터의 날짜 포맷 수정 (ISO 8601)

Commit:
[refactor] 테이블명 컨벤션 통일 (#48)

- invite_links 테이블명을 meeting_invite로 변경

Commit:
[refactor] 테이블명 컨벤션 통일 (#48)

- invite_links 테이블명을 meeting_invite로 변경

Commit:
[refactor] createInviteLink 메서드 @Transactional 제거 (#48)

- 실제 초대 링크를 통해 가입하는 시점에만 트랜잭션 관리하도록 변경

Commit:
[refactor] 초대 코드 생성 로직을 Code 값객체로 분리 (#48)

Commit:
[Refactor] Logger 선언 위치 변경 (#48)

f-lab-saponin reviewed (commented):
PR 반영감사드립니다~! 다시 답변드린부분만 한번 확인부탁드려요. 다음에는 바로 머지가시죠!!

Commit:
[Refactor] 시간 저장 타입을 UTC 기반으로 변경 (#48)

- DATETIME(6)을 TIMESTAMP(6)로 변경
- JPA timezone 설정 추가 (UTC)

f-lab-saponin reviewed (approved):
고생하셨습니다 👍
